tool/src/org/antlr/v4/codegen/model/Lexer.java

```diff
@@ -7,18 +7,19 @@
 package org.antlr.v4.codegen.model;
 
 import org.antlr.v4.codegen.OutputModelFactory;
+import org.antlr.v4.codegen.Target;
 import org.antlr.v4.tool.Grammar;
 import org.antlr.v4.tool.LexerGrammar;
 import org.antlr.v4.tool.Rule;
 
-import java.util.Collection;
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
 
 public class Lexer extends Recognizer {
-	public Map<String,Integer> channels;
-	public LexerFile file;
-	public Collection<String> modes;
+	public final Map<String, Integer> channels;
+	public final Collection<String> channelNames;
```

Comment:
was this field introduced to make code generation easier?

Reply:
Rather more correct. Literal channel and mode names are being preserved, but identifier names are being escaped:

```java
	public static final int
		A=1, B=2;
	public static final int
		break_=2;
	public static final int
		for_=1;
	public static String[] channelNames = {
		"DEFAULT_TOKEN_CHANNEL", "HIDDEN", "break"
	};

	public static String[] modeNames = {
		"DEFAULT_MODE", "for"
	};
```

Reply:
Without looking too closely, I was just wondering why we needed a new field when we must've been generating the list of channel names previously in the code generation.   I tend to avoid saving a field if I can computed on the fly quickly, to prevent having to keep things consistent between fields channels and channelNames, for example.

Reply:
Yes, it's a good note. But we don't have a class for mode or channel that contains information about original and escaped names. Should I introduce them?

```
class Channel {
    int key;
    String name;
    String escapedName;
    ...
```

Reply:
Oh I see.  we do need original and escaped but let's use the HashMap key set for the originals and create escapeChannelNames field for the escaped ones to keep it consistent :)